kedro-telemetry/kedro_telemetry/plugin.py

```diff
@@ -186,40 +195,47 @@ def before_command_run(
                 project_properties, masked_command_args
             )
 
-            _send_heap_event(
-                event_name=f"Command run: {main_command}",
-                identity=user_uuid,
-                properties=cli_properties,
-            )
-
             # send generic event too, so it's easier in data processing
             generic_properties = deepcopy(cli_properties)
             generic_properties["main_command"] = main_command
-            _send_heap_event(
-                event_name="CLI command",
-                identity=user_uuid,
-                properties=generic_properties,
-            )
+            self.event_properties = generic_properties
+
         except Exception as exc:
             logger.warning(
                 "Something went wrong in hook implementation to send command run data to Heap. "
                 "Exception: %s",
                 exc,
             )
 
-
-class KedroTelemetryProjectHooks:
-    """Hook to send project statistics data to Heap"""
+    @cli_hook_impl
+    def after_command_run(self):
```

Comment:
Is this already for the performance issues? #707

Reply:
There are 2 performance gains here:
1. Reduce event 
2. Delay the event during execution to `after_command_run` (side-effect).

The 2 is a co-incident. This is actually carefully designed because I need a unified way to ensure event only sent once for both entrypoints (session.run, kedro run and all other kedro run command).

Requirements:
1. The hook need to be executed in the corresponding entrypoint 
  - `after_command_run` won't be triggered for `session.run`, so it is only related to the `kedro run` flow.
2. Avoid duplicate computation, if the info is collected already, don't collect it again.
3. Send event once and only

Scenario 1 - ` kedro run`
1. before_command_run collect basic info
2. after_catalog_created collect more project statistic (avoid collection basic info twice, i.e. project name)
5. after_command_run send the heap event

Scenario 2 - session.run
1. before_command_run is not triggered
2. after_context_created and after_catalog_created are run, in this case the `after_catalog_created` hook will collect both basic info and project statistic. Event are sent immediately (Maybe can even delay this to after_pipeline_run? See also scenario 4)

Scenario 3 - kedro info / kedro ipython
1. before_command_run collect basic info
2. after_command_run send the actual event

Scenario 4 - kedro catalog create (or other catalog related hook)
Same as scenario 2 because event are sent eagerly after catalog created.